rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -1324,4 +1457,146 @@ size_t Player::get_number_of_registered_on_play_msg_post_callbacks()
+    type == service_msgs::msg::ServiceEventInfo::REQUEST_SENT)
+  {
+    if (message.size() <= rosbag2_cpp::get_serialization_size_for_service_metadata_event()) {
+      client_side_type_ = introspection_type::METADATA;
+      RCLCPP_WARN(
+        logger_,
+        "The configuration of introspection for '%s' client is metadata !",
+        service_name_.c_str());
+      return false;
+    } else {
+      client_side_type_ = introspection_type::CONTENTS;
+    }
+  }
+  if (service_side_type_ == introspection_type::UNKNOW &&
+    type == service_msgs::msg::ServiceEventInfo::REQUEST_RECEIVED)
+  {
+    if (message.size() <= rosbag2_cpp::get_serialization_size_for_service_metadata_event()) {
+      service_side_type_ = introspection_type::METADATA;
+      RCLCPP_WARN(
+        logger_,
+        "The configuration of introspection for '%s' service is metadata !",
+        service_name_.c_str());
+      return false;
+    } else {
+      service_side_type_ = introspection_type::CONTENTS;
+    }
+  }
+
+  // If there are request send info and request receive info, only send request send info.
+  if (client_side_type_ == introspection_type::CONTENTS &&
+    service_side_type_ == introspection_type::CONTENTS &&
+    type == service_msgs::msg::ServiceEventInfo::REQUEST_RECEIVED)
+  {
+    return false;
+  }
+
+  return true;
+}
+
+void PlayerImpl::PlayerClient::async_send_request(const rclcpp::SerializedMessage & message)
+{
+  void * ros_message = new uint8_t[message_members_->size_of_];
+  message_members_->init_function(
+    ros_message, rosidl_runtime_cpp::MessageInitialization::ZERO);
+
+  int ret = rmw_deserialize(&message.get_rcl_serialized_message(), ts_, ros_message);
+  if (ret == RMW_RET_OK) {
+    if (client_->service_is_ready()) {
+      // members_[0]: info, members_[1]: request, members_[2]: response
+      auto request_offset = message_members_->members_[1].offset_;
+      auto request_addr = reinterpret_cast<size_t>(ros_message) + request_offset;
+      client_->async_send_request(
+        reinterpret_cast<void *>(*reinterpret_cast<size_t *>(request_addr)));
+    } else {
+      RCLCPP_ERROR(
+        logger_,
+        "'%s' service isn't ready !",
+        service_name_.c_str());
+    }
+  }
+
+  message_members_->fini_function(ros_message);
+  delete[] static_cast<uint8_t *>(ros_message);
+
+  if (ret != RMW_RET_OK) {
+    throw std::runtime_error(
+            "Failed to deserialize service event message for " + service_name_ + " !");
+  }
+}
+
+uint8_t PlayerImpl::PlayerClient::get_msg_event_type(const rclcpp::SerializedMessage & message)
+{
+  auto msg = service_msgs::msg::ServiceEventInfo();
+
+  const rosidl_message_type_support_t * type_support_info =
+    rosidl_typesupport_cpp::
+    get_message_type_support_handle<service_msgs::msg::ServiceEventInfo>();
+  if (type_support_info == nullptr) {
+    throw std::runtime_error(
+            "It failed to get message type support handle of service event info !");
```

Comment:
Comments from MichaelOrlov

```suggestion
            "Failed to get message type support handle of service event info !");
```

Maybe also add output to the log?

Reply:
Updated